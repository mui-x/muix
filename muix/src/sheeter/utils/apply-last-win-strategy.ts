import { TAtomize } from 'reactxx-typings'
import { globalOptions } from '../global-state'

export interface ApplyLastWinStrategyResult {
    style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
    defferedIdxs?: number[]
}
export const enum AttemptType {
    first,
    firstIgnore,
    second
}

export type ApplyLastWinStrategyLow = (values: TAtomize.AtomicWeb[], attemptType: AttemptType) => ApplyLastWinStrategyResult

export const applyLastwinsStrategyRoot = (values: TAtomize.AtomicArray, applyLastWinStrategyLow: ApplyLastWinStrategyLow) => {
    const processDeffereds = globalOptions.processDeffereds
    let res = applyLastWinStrategyLow(values, processDeffereds ? AttemptType.first : AttemptType.firstIgnore)
    if (res.style) return res.style
    processDeffereds(values, res.defferedIdxs, values.state)
    res = applyLastWinStrategyLow(values, AttemptType.second)
    return res.style
}
