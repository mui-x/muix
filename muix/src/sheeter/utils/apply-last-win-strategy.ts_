import { TAtomize } from 'reactxx-typings'

// export interface ApplyLastWinStrategyResult {
//     style?: TAtomize.AtomicWebsLow | TAtomize.AtomicNativeLow
//     defferedIdxs?: number[]
// }
// export const enum AttemptType {
//     first,
//     firstIgnore,
//     second
// }

