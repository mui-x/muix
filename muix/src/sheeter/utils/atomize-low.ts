import { wrapPseudoPrefixes } from './wrap-pseudo-prefixes'
import { TTyped, TEngine } from 'reactxx-typings'

// platform dependent 
import { platform } from 'reactxx-sheeter'

//*******************************************************
//        adjustAtomizedLow

// linearize and atomize "ruleset"
// result is added to 'atomizedVariants'
export const adjustAtomizedLow = (
    ruleset: TEngine.Rulesets,
    atomizedVariants: TEngine.QueryableItems, path: string,
    pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    processTree(ruleset, atomizedVariants, path, pseudoPrefixes, conditions)
}

export function isToAtomizeArray(obj): obj is TEngine.Ruleset[] { return obj && !(obj as TEngine.Queryables).$r$ && Array.isArray(obj) }
export function isAtomized(obj): obj is TEngine.Queryables { return !obj || (obj as TEngine.Queryables).$r$ }
export function isDeferred(obj): obj is TEngine.Deferred { return obj && (obj as TEngine.Deferred).$d$ }
export function isTemporary(obj): obj is TEngine.TempProc { return obj && (obj as TEngine.TempProc).$t$ }
export function isToAtomize(obj): obj is TEngine.ToAtomize { return obj && !(obj as TEngine.Queryables).$r$ && !(obj as TEngine.Deferred).$d$ && !(obj as TEngine.TempProc).$t$ }

//*******************************************************
//        makeTemporary
export const makeTemporary = (proc: TEngine.TempProc) => {
    proc.$t$ = true
    return proc as any
}

//*******************************************************
//        processTree
export const processTree = (
    value: TEngine.Ruleset | TEngine.Ruleset[], atomizedVariants: TEngine.QueryableItems,
    path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions
) => {
    const items = isToAtomizeArray(value) ? value : [value]
    const indexToPath = (idx: number) => items.length <= 1 ? '' : `[${idx}]`
    items.forEach((it, idx) => {
        if (!it) return
        if (isTemporary(it)) {
            it(atomizedVariants, path + indexToPath(idx), pseudoPrefixes, conditions)
        } else if (isDeferred(it)) {
            atomizedVariants.push(it)
        } else if (isAtomized(it)) {
            if (pseudoPrefixes && pseudoPrefixes.length > 0)
                throw 'Web pseudo properties cannot contain already atomized value'
            it.forEach(it => {
                if (!conditions || conditions.length === 0) //... and no conditions => done
                    pushToAtomizedVariants(atomizedVariants, it, null)
                else { // ... and conditions => merge conditions
                    pushToAtomizedVariants(atomizedVariants, [...it],
                        it.conditions ? [...conditions, ...it.conditions] : conditions)
                }
            })
        } else { // it: toAtomize
            processPseudosAndAtomize(
                it,
                atomizedVariants, path + indexToPath(idx),
                pseudoPrefixes, conditions
            )
        }
    })
}

//*********************************************************
//  PRIVATE
//*********************************************************

//*******************************************************
//        processAllInnerVariants

// recursivelly process web pseudos (:hover, :active) and its queryable parts

const processPseudosAndAtomize = (
    ruleset: TEngine.ToAtomize,
    atomizedVariants: TEngine.QueryableItems, path: string, pseudoPrefixes: string[], conditions: TEngine.Conditions,
) => {

    const inner: TEngine.QueryableItems = []

    // process and remove conditional parts of self and web pseudo (e.g. :hover etc.)
    // push them to 'inner'
    for (const p in ruleset) {
        const value = ruleset[p] as TEngine.ToAtomize
        if (!value || p.charAt(0) === '$' || (typeof value !== 'object' && !isTemporary(value))) continue
        processTree(value, inner, `${path}/${p}`, [...pseudoPrefixes, p], conditions)
        delete ruleset[p]
    }

    // atomize pure ruleset tree with pseudo (and without all conditions, deffer's etc.),
    // then push to result
    atomizeNonVariant(atomizedVariants, wrapPseudoPrefixes(ruleset, pseudoPrefixes), conditions, path)

    // push 'inner' to result
    if (inner.length > 0)
        Array.prototype.push.apply(atomizedVariants, inner)

}

// *********** utils

const pushToAtomizedVariants = (
    atomizedVariants: TEngine.QueryableItems, 
    variant: TEngine.Queryable, conditions: TEngine.Conditions
    ) => {
    if (!variant) return
    if (conditions && conditions.length > 0) variant.conditions = conditions
    atomizedVariants.push(variant)
}

const atomizeNonVariant = (
    atomizedVariants: TEngine.QueryableItems, sourceRuleset: TEngine.Rulesets,
    conditions: TEngine.Conditions, path: string
) => {
    if (!sourceRuleset) return
    if (isDeferred(sourceRuleset)) {
        atomizedVariants.push(sourceRuleset)
        return
    }
    // platform specific atomize: e.g. for WEB convert ruleset to list of atomized CSS classes
    const variant = platform.toPlatformAtomizeRuleset(sourceRuleset, path)
    pushToAtomizedVariants(atomizedVariants, variant, conditions)
}

