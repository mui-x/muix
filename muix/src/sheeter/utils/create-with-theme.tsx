import warning from 'warning'

export const createWithTheme = (valueOrCreator: ((theme) => any) | any, theme) => {
  if (typeof valueOrCreator === 'function') {
    warning(theme, 'Theme expected (ThemeProvider or getDefaultTheme missing)')
    // apply creator to theme:
    return valueOrCreator(theme)
  }
  else
    // return value
    return valueOrCreator
};