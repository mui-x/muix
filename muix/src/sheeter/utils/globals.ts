import { TExtensions, TEngine } from 'reactxx-typings';
import { WidthStore } from '../queryable/$widths/store';

export interface Platform extends TExtensions.Platform {
    toPlatformAtomizeRuleset?: TEngine.ToPlatformAtomizeRuleset
    dataTrace?: (classNames: TEngine.AtomicArrayLow, flags?: TraceFlags) => any,
    applyLastwinsStrategy?: TEngine.ApplyLastwinsStrategy
    finalizeClassName?: (values: TEngine.AtomicArrayLow) => string | Record<string, any>
    getDefaultTheme?: () => any
}

export const resetPlatform = () => {
    for (const p in platform) delete platform[p]
}

export const assignPlatform = (gl: Platform) => Object.assign(platform, gl)

export const platform: Platform = {}

export const initGlobals = (force: boolean, platformDependentInit: () => void) => {
    if (force) resetPlatform()
    if (platform._sheeter) return

    platform._sheeter = {}
    // fill e.g. actWidth
    platformDependentInit()
    // init store with act width
    platform._sheeter.widthsStore = new WidthStore()

}