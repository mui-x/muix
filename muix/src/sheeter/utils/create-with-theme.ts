import warning from 'warning'
import { isTemporary } from './atomize-low'

export const createWithTheme = (valueOrCreator: ((theme) => any) | any, theme) => {
  if (typeof valueOrCreator === 'function' && !isTemporary(valueOrCreator))
    // apply creator to theme:
    return valueOrCreator(theme)
  else // return value
    return valueOrCreator
}