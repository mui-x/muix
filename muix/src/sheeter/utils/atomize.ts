import warning from 'warning';

import { TEngine, TTyped, TSheeter, TComponents, TCommonStyles } from 'reactxx-typings'
import { createWithTheme } from './create-with-theme'
import { adjustAtomizedLow, isDeferred } from './atomize-low'

// muttable
export const atomizeSheet = (sheet: TEngine.SheetOrCreator, theme?, path: string = 'sheet') => {
    if (!sheet) return null
    const sh: TEngine.Sheet = createWithTheme(sheet, theme)
    for (const p in sh) {
        const at = atomizeRuleset(sh[p], theme, (path ? path + '.' : '') + p)
        if (at) sh[p] = at
        else delete sh[p]
    }
    return sh
}

// muttable
export const atomizeRuleset = (
    ruleset: TEngine.RulesetOrCreator, theme?, path: string = ''
) => {
    if (!ruleset) return null

    const rs = createWithTheme(ruleset, theme)

    if (!rs) return null

    const list = wrapRuleset([])
    adjustAtomizedLow(rs, list, path, [], [])

    return list.length === 0 ? null : list
}

export const wrapRuleset = ruleset => {
    (ruleset as TEngine.Queryables).$r$ = true
    if (window.__TRACE__)
        ruleset['toJSON'] = toJSON.bind(ruleset)
    return ruleset as TEngine.Queryables
}
function toJSON () {
    return (this as TEngine.Queryables).map(v => isDeferred(v) ? 'DEFFERED' : [...v])
}

// muttable (at least for native)
export const atomizeStyle = (style: TSheeter.StyleOrCreator, theme?, path: string = '.') => {
    if (!style) return null
    if (window.isWeb)
        return createWithTheme(style, theme) as TSheeter.StyleOrAtomizedWeb
    else
        return atomizeRuleset(style as TEngine.Rulesets, theme, path)
}

export function isReactXXComponent(obj): obj is TComponents.ComponentType {
    return (obj as TComponents.ComponentType).$c$
}