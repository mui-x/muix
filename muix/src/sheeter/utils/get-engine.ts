import { TTyped } from 'reactxx-typings';
import { $width, $web, $native, $hot, $if, $ifelse } from '../queryable/index'
import { toClassNamesWithQuery } from '../utils/to-classnames'
import { mergeRulesets, mergeSheets } from '../utils/merge'
import { atomizeRuleset, atomizeSheet } from '../utils/atomize'

export const getEngine = <R extends TTyped.Shape>() => untypedEngine as TTyped.TypedEngine<R>

const untypedEngine = {
    WEB: $web,
    NATIVE: $native,
    HOT: $hot,
    IF: $if,
    IFELSE: $ifelse,
    WIDTH: $width,
    THEMED: p => p,
    STYLE: (...p) => p,
    $atomizeSheet: atomizeSheet as any,
    $mergeSheets: mergeSheets as any,
    $atomizeRuleset: atomizeRuleset as any,
    ATOMIZE: (...pars: any[]) => atomizeRuleset(pars, null, '$atomize'),
    $mergeRulesets: mergeRulesets as any,
    $toClassNames: toClassNamesWithQuery as any
} as TTyped.TypedEngine<TTyped.Shape>
