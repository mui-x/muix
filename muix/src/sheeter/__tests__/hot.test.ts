import { $W, $T, $V, $I, V, T, I, TTyped } from 'reactxx-typings'
import { getEngine } from '../utils/get-engine'

//let $atomizeRuleset, $toClassNames
import { initPlatform, dump, afterLastWin } from "./init-platform.t"

import { theme, Theme } from "reactxx-typings-test/shape.t"

interface Shape {
  theme: Theme
  sheetQuery: { primary: string }
}

const { $themed, $if, $web, $hot, $native, $rules, $toClassNames, $atomizeRuleset, $ifelse, $mergeRulesets, $width
} = getEngine<Shape>()


describe("SHEETER HOT", () => {

  const doTest = (isWeb: boolean) => {
    beforeEach(() => initPlatform(isWeb))
    let ruleset

    it("01 null", () => {
      ruleset = $atomizeRuleset<V>([
        $hot<V>(null),
      ])
      expect(ruleset).toMatchSnapshot()
    })

    it("02 empty", () => {
      ruleset = $atomizeRuleset([
        $hot(null),
        $hot(() => null),
        $hot(() => [null, {}, undefined]),
      ])
      expect(ruleset).toMatchSnapshot()
    })

    it("03 just $hot", () => {
      ruleset = $atomizeRuleset([
        { color: 'green' },
        $hot(state => ({
          color: 'red'
        })),
      ])
      expect(ruleset).toMatchSnapshot()
    })

    it("04: color: 'red'", () => afterLastWin($toClassNames<T>(null,
      $hot<T>(state => ({
        color: 'red'
      }))
    )))

    it("05: nested", () => afterLastWin($toClassNames<T>(null,
      $hot<T>(state => $hot<T>(state => ({ color: 'red' })))
    )))

    it("06: nested", () => afterLastWin($toClassNames<T>(null,
      $hot<T>(state => [{ margin: 0 }, $hot<T>(state => ({ color: 'red' }))])
    )))

    describe("07: use attr", () => {
      beforeEach(() => initPlatform(isWeb))

      // '{ color: queryPar.primary }' is processet very late (during $toClassNames processing)
      const RS1 = $hot<T>(queryPar => ({ color: queryPar.$sheetQuery.primary }))
      const RS2 = $hot<T>(queryPar => $if<T>(true, { color: queryPar.$sheetQuery.primary }))
      const RS3 = $hot<T>(queryPar => [
        $if<T>(queryPar => queryPar.$sheetQuery.primary === 'red', { color: queryPar.$sheetQuery.primary }),
        $hot<T>(queryPar => ({ backgroundColor: queryPar.$sheetQuery.primary }))
      ])

      const test = (msg: string, ruleset: T) => {
        it("01: red " + msg, () => afterLastWin($toClassNames<T>({ $sheetQuery: { primary: 'red' } },
          ruleset // color is red
        )))
        it("02: green " + msg, () => afterLastWin($toClassNames<T>({ $sheetQuery: { primary: 'green' } },
          ruleset // color is green
        )))
      }
      test('RS1', RS1)
      test('RS2', RS2)
      test('RS3', RS3)
    })

  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})
