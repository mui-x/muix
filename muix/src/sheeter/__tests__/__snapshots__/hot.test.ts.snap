// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHEETER HOT ## NATIVE ## 01 null 1`] = `
Array [
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## NATIVE ## 02 empty 1`] = `
Array [
  "DEFFERED",
  "DEFFERED",
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## NATIVE ## 03 just $hot 1`] = `
Array [
  Object {
    "color": "green @[0]",
  },
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## NATIVE ## 04: color: 'red' 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot",
    "value": "red",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 05: nested 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot/$hot",
    "value": "red",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 06: nested 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot[1]/$hot",
    "value": "red",
  },
  "margin": Object {
    "tracePath": "/$hot[0]",
    "value": 0,
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 01: red RS1 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot",
    "value": "red",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 01: red RS2 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot/$sif",
    "value": "red",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 01: red RS3 1`] = `
Object {
  "backgroundColor": Object {
    "tracePath": "/$hot[1]/$hot",
    "value": "red",
  },
  "color": Object {
    "tracePath": "/$hot[0]/$if",
    "value": "red",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 02: green RS1 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot",
    "value": "green",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 02: green RS2 1`] = `
Object {
  "color": Object {
    "tracePath": "/$hot/$sif",
    "value": "green",
  },
}
`;

exports[`SHEETER HOT ## NATIVE ## 07: use attr 02: green RS3 1`] = `
Object {
  "backgroundColor": Object {
    "tracePath": "/$hot[1]/$hot",
    "value": "green",
  },
}
`;

exports[`SHEETER HOT ## WEB ## 01 null 1`] = `
Array [
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## WEB ## 02 empty 1`] = `
Array [
  "DEFFERED",
  "DEFFERED",
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## WEB ## 03 just $hot 1`] = `
Array [
  Array [
    ".a { color:green } @[0]",
  ],
  "DEFFERED",
]
`;

exports[`SHEETER HOT ## WEB ## 04: color: 'red' 1`] = `
Array [
  ".a { color:red } @/$hot",
]
`;

exports[`SHEETER HOT ## WEB ## 05: nested 1`] = `
Array [
  ".a { color:red } @/$hot/$hot",
]
`;

exports[`SHEETER HOT ## WEB ## 06: nested 1`] = `
Array [
  ".b { color:red } @/$hot[1]/$hot",
  ".a { margin:0 } @/$hot[0]",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 01: red RS1 1`] = `
Array [
  ".a { color:red } @/$hot",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 01: red RS2 1`] = `
Array [
  ".a { color:red } @/$hot/$sif",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 01: red RS3 1`] = `
Array [
  ".b { background-color:red } @/$hot[1]/$hot",
  ".a { color:red } @/$hot[0]/$if",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 02: green RS1 1`] = `
Array [
  ".a { color:green } @/$hot",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 02: green RS2 1`] = `
Array [
  ".a { color:green } @/$hot/$sif",
]
`;

exports[`SHEETER HOT ## WEB ## 07: use attr 02: green RS3 1`] = `
Array [
  ".b { background-color:green } @/$hot[1]/$hot",
]
`;
